eva_submission/biosamples_submission.py

```diff
@@ -171,26 +181,63 @@ def validate_in_bsd(self, samples_data):
         for sample in samples_data:
             # If we're only retrieving, don't need to validate.
             if not self.should_retrieve(sample):
-                sample['domain'] = self.domain
+                if hasattr(self.communicator, "domain"):
+                    sample['domain'] = self.communicator.domain
+                if hasattr(self.communicator, "webin_submission_account_id"):
+                    sample['webinSubmissionAccountId'] = self.communicator.webin_submission_account_id
                 self.communicator.follows_link('samples', join_url='validate', method='POST', json=sample)
 
+    def convert_sample_data_to_curation_object(self, future_sample):
+        """Curation object can only change 3 attributes characteristics, externalReferences and relationships"""
+        current_sample = self.communicator.follows_link('samples', method='GET', join_url=future_sample.get('accession'))
+        curation_object = {}
+        attributes_pre = []
+        attributes_post = []
+        for attribute in future_sample['characteristics']:
+            if attribute in current_sample['characteristics']:
+                attributes_pre.append({
+                    'type': attribute,
+                    'value': current_sample['characteristics'].get(attribute)[0].get('text')
+                })
+            # else:
+            #     attributes_pre.append({})
+            attributes_post.append({
+                'type': attribute,
+                'value': future_sample['characteristics'].get(attribute)[0].get('text')
+            })
+        curation_object['attributesPre'] = attributes_pre
+        curation_object['attributesPost'] = attributes_post
+
+        # for externalReferences and relationships we're assuming you want to replace them all
+        curation_object['externalReferencesPre'] = current_sample.get('externalReferences', [])
+        curation_object['externalReferencesPost'] = future_sample.get('externalReferences', [])
+        curation_object['relationshipsPre'] = current_sample.get('relationships', [])
+        curation_object['relationshipsPost'] = future_sample.get('relationships', [])
+
+        return dict(curation=curation_object, sample=future_sample.get('accession'))
+
     def submit_to_bsd(self, samples_data):
         """
         This function creates or updates samples in BioSamples and return a map of sample name to sample accession
         """
         for sample in samples_data:
-            sample['domain'] = self.domain
+            if hasattr(self.communicator, "domain"):
+                sample['domain'] = self.communicator.domain
+            if hasattr(self.communicator, "webin_submission_account_id"):
+                sample['webinSubmissionAccountId'] = self.communicator.webin_submission_account_id
```

Comment:
Is there a nice way to abstract this, since it's now slightly fiddly and repeats in lines 184-187?  Maybe just an `add_communicator_attributes(sample)` method or something, to modify the dict with anything that is needed for the biosamples communication but isn't really part of the sample.

Reply:
That's a good point. I've refactored that part of the code.